core/include/traccc/clusterization/impl/measurement_creation.ipp

```diff
@@ -92,38 +100,50 @@ TRACCC_HOST_DEVICE inline void fill_measurement(
     // [2] The Art of Computer Programming, Donald E. Knuth, second
     //     edition, chapter 4.2.2.
 
-    // A security check.
-    assert(cluster.empty() == false);
+    // Security checks.
+    assert(clusters.cell_indices()[index].empty() == false);
+    assert([&]() {
+        const unsigned int module_idx =
+            cells.module_index().at(clusters.cell_indices()[index].front());
+        for (const unsigned int cell_idx : clusters.cell_indices()[index]) {
+            if (cells.module_index().at(cell_idx) != module_idx) {
+                return false;
+            }
+        }
+        return true;
+    }() == true);
```

Comment:
```suggestion
    assert(std::all_of(clusters.cell_indices()[index].cbegin(), clusters.cell_indices()[index].cend(), [module_idx=cells.module_index().at(clusters.cell_indices()[index].front())](module_idx_type & m){ m == module_idx}));
```

Reply:
I'm afraid of using `std::all_of` here... :thinking: This function advertises itself as working in device code. And the STL algorithms don't have a terrific track record with that.

But I very much agree, I'm not in love with this assertion at the moment either. :frowning:

Reply:
Perhaps we could implement out own `all_of`? That we can guarantee will run on device?

Reply:
I guess. Though as long as we only have a single use, that seems like a waste of effort. I would introduce that function once we have a second client. :thinking:

Reply:
Okay, that's fair. :+1: